### PR TITLE
clientv3: fix init client error

### LINKED ISSUE #14323: Etcd v3 api: Initialization client error
### What happened?

etcd client v3:  v3.5.4

I get an error when I initialize the client.

When I change `RejectOldCluster` to false everything works fine.
```go
	cfg := clientv3.Config{
		Endpoints:            []string{"192.168.3.41:22479", "192.168.3.41:22379", "192.168.3.41:22579"},
		Username:             "root",
		Password:             "123456",
		RejectOldCluster:     true,
		DialTimeout:          5 * time.Second,
		DialKeepAliveTime:    5 * time.Second,
		DialKeepAliveTimeout: 5 * time.Second,
		AutoSyncInterval:     time.Minute,
	}
```

```shell
{"level":"warn","ts":"2022-08-08T16:05:33.632+0800","logger":"etcd-client","caller":"v3@v3.5.4/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001fb500/192.168.3.41:22479","attempt":0,"error":"rpc error: code
 = Canceled desc = context canceled"}
{"level":"warn","ts":"2022-08-08T16:05:33.639+0800","logger":"etcd-client","caller":"v3@v3.5.4/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001fb500/192.168.3.41:22479","attempt":0,"error":"rpc error: code
 = Canceled desc = context canceled"}

```

There is a problem at line 512 of `client.go` where it should not break when there are no err.
If other goroutines are slow, context canceled will occur.
But this problem does not affect normal use.
image

### What did you expect to happen?

No error output.

### How can we reproduce it (as minimally and precisely as possible)?

_step 1_

```shell
docker-compose up -d
```

`docker-compose.yaml`:
```yaml
version: "3.0"
services:
  etcd1:
    image: quay.io/coreos/etcd:v3.5.4
    container_name: "etcd1"
    networks:
      internal:
        ipv4_address: 172.25.1.3
    ports:
      - "22379-22380:22379-22380"
    volumes:
      - ./etcd1:/var/etcd/etcd-data
    environment:
      - ETCD_NAME=etcd1
      - ETCD_DATA_DIR=/var/etcd/etcd-data
      - ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:22379
      - ETCD_LISTEN_PEER_URLS=http://0.0.0.0:22380
      - ETCD_ADVERTISE_CLIENT_URLS=http://192.168.3.41:22379
      - ETCD_INITIAL_ADVERTISE_PEER_URLS=http://172.25.1.3:22380
      - ETCD_INITIAL_CLUSTER_TOKEN=etcd-cluster
      - ETCD_INITIAL_CLUSTER=etcd1=http://172.25.1.3:22380,etcd2=http://172.25.1.4:22480,etcd3=http://172.25.1.5:22580
      - ETCD_INITIAL_CLUSTER_STATE=new      
  etcd2:
    image: quay.io/coreos/etcd:v3.5.4
    container_name: "etcd2"
    networks:
      internal:
        ipv4_address: 172.25.1.4
    ports:
      - "22479-22480:22479-22480"
    volumes:
      - ./etcd2:/var/etcd/etcd-data
    environment:
      - ETCD_NAME=etcd2
      - ETCD_DATA_DIR=/var/etcd/etcd-data
      - ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:22479
      - ETCD_LISTEN_PEER_URLS=http://0.0.0.0:22480
      - ETCD_ADVERTISE_CLIENT_URLS=http://192.168.3.41:22479
      - ETCD_INITIAL_ADVERTISE_PEER_URLS=http://172.25.1.4:22480
      - ETCD_INITIAL_CLUSTER_TOKEN=etcd-cluster
      - ETCD_INITIAL_CLUSTER=etcd1=http://172.25.1.3:22380,etcd2=http://172.25.1.4:22480,etcd3=http://172.25.1.5:22580
      - ETCD_INITIAL_CLUSTER_STATE=new      
  etcd3:
    image: quay.io/coreos/etcd:v3.5.4
    container_name: "etcd3"
    networks:
      internal:
        ipv4_address: 172.25.1.5
    ports:
      - "22579-22580:22579-22580"
    volumes:
      - ./etcd3:/var/etcd/etcd-data
    environment:
      - ETCD_NAME=etcd3
      - ETCD_DATA_DIR=/var/etcd/etcd-data
      - ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:22579
      - ETCD_LISTEN_PEER_URLS=http://0.0.0.0:22580
      - ETCD_ADVERTISE_CLIENT_URLS=http://192.168.3.41:22579
      - ETCD_INITIAL_ADVERTISE_PEER_URLS=http://172.25.1.5:22580
      - ETCD_INITIAL_CLUSTER_TOKEN=etcd-cluster
      - ETCD_INITIAL_CLUSTER=etcd1=http://172.25.1.3:22380,etcd2=http://172.25.1.4:22480,etcd3=http://172.25.1.5:22580
      - ETCD_INITIAL_CLUSTER_STATE=new       
networks:
  internal:
    driver: bridge
    ipam:
      config:
        - subnet: 172.25.1.0/16
```
_step 2_

Initialize client.

```go
	cfg := clientv3.Config{
		Endpoints:            []string{"192.168.3.41:22479", "192.168.3.41:22379", "192.168.3.41:22579"},
		RejectOldCluster:     true,
		DialTimeout:          5 * time.Second,
		DialKeepAliveTime:    5 * time.Second,
		DialKeepAliveTimeout: 5 * time.Second,
		AutoSyncInterval:     time.Minute,
	}
       client, err := clientv3.New(cfg)
```

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
# paste output here
etcd Version: 3.5.4
Git SHA: 08407ff76
Go Version: go1.16.15
Go OS/Arch: linux/amd64

$ etcdctl version
# paste output here
etcdctl version: 3.5.4
API version: 3.5
```

</details>

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands blow, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

+------------------+---------+-------+--------------------------+----------------------------+------------+
|        ID        | STATUS  | NAME  |        PEER ADDRS        |        CLIENT ADDRS        | IS LEARNER |
+------------------+---------+-------+--------------------------+----------------------------+------------+
|  4aea8f1af490926 | started | etcd1 | https://172.25.1.3:22380 | https://192.168.3.41:22379 |      false |
| 43038eadf883f7e2 | started | etcd3 | https://172.25.1.5:22580 | https://192.168.3.41:22579 |      false |
| 7191a6e6cc023b1a | started | etcd2 | https://172.25.1.4:22480 | https://192.168.3.41:22479 |      false |
+------------------+---------+-------+--------------------------+----------------------------+------------+

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
+----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
|          ENDPOINT          |        ID        | VERSION | DB SIZE | IS LEADER | IS LEARNER | RAFT TERM | RAFT INDEX | RAFT APPLIED INDEX | ERRORS |
+----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+
| https://192.168.3.41:22379 |  4aea8f1af490926 |   3.5.4 |   20 kB |     false |      false |         3 |         42 |                 42 |        |
| https://192.168.3.41:22479 | 7191a6e6cc023b1a |   3.5.4 |   20 kB |     false |      false |         3 |         42 |                 42 |        |
| https://192.168.3.41:22579 | 43038eadf883f7e2 |   3.5.4 |   20 kB |      true |      false |         3 |         42 |                 42 |        |
+----------------------------+------------------+---------+---------+-----------+------------+-----------+------------+--------------------+--------+

```

</details>

### Relevant log output

```shell
{"level":"warn","ts":"2022-08-08T16:05:33.632+0800","logger":"etcd-client","caller":"v3@v3.5.4/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001fb500/192.168.3.41:22479","attempt":0,"error":"rpc error: code
 = Canceled desc = context canceled"}
{"level":"warn","ts":"2022-08-08T16:05:33.639+0800","logger":"etcd-client","caller":"v3@v3.5.4/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0001fb500/192.168.3.41:22479","attempt":0,"error":"rpc error: code
 = Canceled desc = context canceled"}
```

### PR DESCRIPTION
**Pull request by happlins**
fix clientv3 init client error

close #14323

Signed-off-by: happlins [happlins@foxmail.com](mailto:happlins@foxmail.com)

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by ahrtr** [comment]
Two comments:
1. please  signoff the commit using command `git rebase HEAD~1 --signoff`;
2. Please consider to add a couple of unit tests to verify the changes.

**Comment by happlins** [comment]
I have modified and added unit tests.

**Comment by SimFG** [comment]
@happlins you need to combine four commits into one

**Comment by happlins** [comment]
sorry,I have revised it.

File: client/v3/client_test.go [hunk 1]
```diff
@@ -266,6 +266,29 @@ func TestSyncFiltersMembers(t *testing.T) {
...
+func TestMultiEndpoints(t *testing.T) {
+	cfg := Config{
+		Endpoints:            []string{"192.168.3.41:22379", "192.168.3.41:22479", "192.168.3.41:22579"},
+		RejectOldCluster:     true,
+		DialTimeout:          5 * time.Second,
+		DialKeepAliveTime:    5 * time.Second,
+		DialKeepAliveTimeout: 5 * time.Second,
+		AutoSyncInterval:     time.Minute,
+	}
+	c, err := NewClient(t, cfg)
```
Comment (ptabor) [inline client/v3/client_test.go:278]: The client is not `defer` Closed... thus it leaks goroutines and causes the tests to fail. 

(The issue would be easier to catch/debug if all the tests had: 
`testutil.BeforeTest(t)` as a first line). Would you mind adding in a separate PR ?

Comment (happlins) [reply client/v3/client_test.go:279]: I have revised it.

**Comment by ahrtr** [comment]
The minimum version supported should be `3.4`, so probably we should update `min` at [client.go#L510](https://github.com/etcd-io/etcd/blob/4d57eb8d3b2190bfdba3c65f8eb93c0349fc6dcc/client/v3/client.go#L510).

Please also consider to enhance the unit test to cover more scenarios,such as,
1. All endpoint versions are up to date;
2. All endpoint versions are out of date;
3. Partially endpoint versions are out of date;

**Comment by ahrtr** [comment]
any update on this?

**Comment by happlins** [comment]
I'm sorry. I've been too busy recently. I've modified the minimum version and added unit test

File: client/v3/client_test.go [hunk 2]
```diff
@@ -266,6 +268,106 @@ func TestSyncFiltersMembers(t *testing.T) {
...
+				ctx:       context.Background(),
+				endpoints: []string{"192.168.3.41:22379", "192.168.3.41:22479", "192.168.3.41:22579"},
+				epMu:      new(sync.RWMutex),
+				Maintenance: &mockMaintenance{
+					Version: map[string]string{"192.168.3.41:22379": "3.3.0", "192.168.3.41:22479": "3.3.0", "192.168.3.41:22579": "3.3.0"},
+				},
+			},
+			err: ErrOldCluster,
+		},
+	}
```
Comment (ahrtr) [inline client/v3/client_test.go:321]:
```suggestion
	var tests = []struct {
		name              string
		endpoints       []string
		versions          []string
		expectedError error
	}{
		{
			name: "all new versions with the same value",
			......
			err: nil,
		},
		{
			name: "all new versions with different values",
			......
			err: nil,
		},
		{
			name: "all old versions with different values",
			......
			err: ErrOldCluster,
		},
		{
			name: "all old versions with the same value",
			......
			err: ErrOldCluster,
		},
	}
```

Comment (happlins) [reply client/v3/client_test.go]: Thanks, it has been modified.

File: client/v3/client_test.go [hunk 3]
```diff
@@ -266,6 +268,105 @@ func TestSyncFiltersMembers(t *testing.T) {
...
+		},
+		{
+			name:          "all old versions with the same value",
+			endpoints:     []string{"192.168.3.41:22379", "192.168.3.41:22479", "192.168.3.41:22579"},
+			versions:      []string{"3.3.0", "3.3.0", "3.3.0"},
+			expectedError: ErrOldCluster,
+		},
+	}
+
+	for i, tt := range tests {
```
Comment (ahrtr) [inline client/v3/client_test.go:305]: Please use golang subtest something like below,
```
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
                    ......
		})
	}
```

Comment (serathius) [reply client/v3/client_test.go:305]: Using `t.Run` will make it a separate test, which means no need to use `i` in logging to identify test case.

Comment (happlins) [reply client/v3/client_test.go]: I have modified.

File: client/v3/client_test.go [hunk 4]
```diff
@@ -266,6 +268,108 @@ func TestSyncFiltersMembers(t *testing.T) {
...
+			endpoints:     []string{"192.168.3.41:22379", "192.168.3.41:22479", "192.168.3.41:22579"},
+			versions:      []string{"3.3.0", "3.3.0", "3.3.0"},
+			expectedError: ErrOldCluster,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if len(tt.endpoints) != len(tt.versions) || len(tt.endpoints) == 0 {
+				t.Errorf(" endpoints & version length err")
```
Comment (ahrtr) [inline client/v3/client_test.go:308]: There is no any variadic parameters, so you shouldn't use `Errorf`. 

```suggestion
				t.Errorf("Unexpected endpoints and versions length, len(endpoints):%d, len(versions):%d", len(endpoints), len(versions))
```

**Review by ahrtr — APPROVED** [review]
LGMT

Thank you @happlins

**Comment by ahrtr** [comment]
Since there is no response from other maintainers and it's a safe change to me, so let me merge this PR.

**Pull request merged**